events/service/src/main/java/org/projectnessie/events/service/EventService.java

```diff
@@ -0,0 +1,303 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.events.service;
+
+import jakarta.annotation.Nullable;
+import java.security.Principal;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import org.projectnessie.events.api.CommitEvent;
+import org.projectnessie.events.api.Content;
+import org.projectnessie.events.api.ContentKey;
+import org.projectnessie.events.api.ContentRemovedEvent;
+import org.projectnessie.events.api.ContentStoredEvent;
+import org.projectnessie.events.api.Event;
+import org.projectnessie.events.api.EventType;
+import org.projectnessie.events.api.MergeEvent;
+import org.projectnessie.events.api.ReferenceCreatedEvent;
+import org.projectnessie.events.api.ReferenceDeletedEvent;
+import org.projectnessie.events.api.ReferenceUpdatedEvent;
+import org.projectnessie.events.api.TransplantEvent;
+import org.projectnessie.events.service.util.ContentMapping;
+import org.projectnessie.events.spi.EventSubscriber;
+import org.projectnessie.events.spi.EventSubscription;
+import org.projectnessie.events.spi.ImmutableEventSubscription;
+import org.projectnessie.versioned.BranchName;
+import org.projectnessie.versioned.Commit;
+import org.projectnessie.versioned.CommitResult;
+import org.projectnessie.versioned.Hash;
+import org.projectnessie.versioned.MergeResult;
+import org.projectnessie.versioned.Operation;
+import org.projectnessie.versioned.Put;
+import org.projectnessie.versioned.ReferenceAssignedResult;
+import org.projectnessie.versioned.ReferenceCreatedResult;
+import org.projectnessie.versioned.ReferenceDeletedResult;
+import org.projectnessie.versioned.Result;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.MDC;
+
+/**
+ * Base class for event services.
+ *
+ * <p>This class takes care of starting and stopping the subscribers, and provides helper methods
+ * for handling incoming results from the version store, and for firing events.
+ *
+ * <p>This class is meant to be used as a singleton. It provides all the required functionality to
+ * process and deliver events. Subclasses may override some of the protected methods to add support
+ * for tracing, or to implement more sophisticated delivery logic.
+ */
+public class EventService implements AutoCloseable {
+
+  public static final String SUBSCRIPTION_ID_MDC_KEY = "nessie.events.subscription.id";
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(EventService.class);
+
+  protected final EventConfig config;
+  protected final EventFactory factory;
+  protected final EventSubscribers subscribers;
+  private final Map<EventSubscription, EventSubscriber> subscriptions = new ConcurrentHashMap<>();
```

Comment:
Not sure that we need a CHM here - it might only be necessary during startup, but not during runtime.

Should we build a non-concurrent map during `start()` and just initialize this field when done?